### PR TITLE
feat: attention for skinny Q tensor

### PR DESCRIPTION
**Pull request by pommedeterresautee**
related to #153 

During the writing of the kernel, it appears its accesses were not optimized.
https://github.com/openai/triton/issues/949

```log
❯ pytest test/test_torchdynamo.py -k "benchmark and optimized" -x       
====================================================================================== 104 passed, 162 deselected in 2035.37s (0:33:55) =======================================================================================

```

### CODE REVIEW

**Comment by pommedeterresautee** [comment]
test pass
```

============================================================================================== 2835 passed in 8631.67s (2:23:51) ==============================================================================================


```

File: src/kernl/implementations/cuda_graph.py [hunk 1]
```diff
@@ -41,6 +41,8 @@ def cuda_graphs_wrapper(
     stream.synchronize()
     torch.cuda.current_stream().wait_stream(stream)
     torch.cuda.synchronize()
+    # copy inputs after executing the warmup in case it mutates them at the first iteration
+    static_inputs = [torch.zeros_like(x) for x in inputs]
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/implementations/cuda_graph.py:45]: during the very first warmup, the input layout of V tensor is modified to change its strides.
We need to perform the copy after the change, so after the warmup.

File: src/kernl/implementations/attention_vec_mat.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    sm_scale: float,
+    is_causal: bool = False,
+    attention_mask: None = None,
+):
+    assert is_causal is False, "causal mask is not supported"
+    assert attention_mask is None, "attention_mask is not supported"
+    assert output.shape == q.shape[:3] + k.shape[-1:], f"{output.shape} != {q.shape[:3] + k.shape[-1:]}"
+    if v.stride()[-1] == 1:  # is row major
+        # change layout to col major
+        v.set_(source=v.permute(0, 1, 3, 2).contiguous().permute(0, 1, 3, 2))
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/implementations/attention_vec_mat.py:168]: done

File: src/kernl/implementations/attention_vec_mat.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    out_ptr = out_tensor + (
+        batch_idx * out_batch_stride
+        + n_head_idx * out_n_head_stride
+        + (n_block_idx * SIZE_N + size_n_arange) * out_cols_stride
+    )
+    out_mask = (n_block_idx * SIZE_N + size_n_arange) < out_cols
+    tl.store(pointer=out_ptr, value=result, mask=out_mask, eviction_policy="evict_first")
+
+
+def vec_mat_wrapper(
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/implementations/attention_vec_mat.py:90]: done

File: src/kernl/implementations/attention_vec_mat.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    matrix_batch_stride,
+    matrix_n_head_stride,
+    matrix_rows_stride,
+    matrix_cols_stride,
+    out_batch_stride,
+    out_n_head_stride,
+    out_rows_stride,
+    out_cols_stride,
+    SCALER: tl.constexpr,
+    VEC_SOFTMAX: tl.constexpr,
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/implementations/attention_vec_mat.py:44]: done

File: src/kernl/implementations/attention_vec_mat.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        triton.Config({"SIZE_N": 8}, num_warps=1, num_stages=1),
+        triton.Config({"SIZE_N": 4}, num_warps=1, num_stages=1),
+        triton.Config({"SIZE_N": 2}, num_warps=1, num_stages=1),
+        triton.Config({"SIZE_N": 1}, num_warps=1, num_stages=1),
+    ],
+    key=["vec_cols", "mat_rows", "mat_cols", "out_cols"],
+)
+@triton.jit
+def vec_mat(
+    vec_cols: tl.constexpr,
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/implementations/attention_vec_mat.py:24]: done

File: src/kernl/implementations/attention_vec_mat.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        triton.Config({"SIZE_N": 16}, num_warps=1, num_stages=1),
+        triton.Config({"SIZE_N": 8}, num_warps=1, num_stages=1),
+        triton.Config({"SIZE_N": 4}, num_warps=1, num_stages=1),
+        triton.Config({"SIZE_N": 2}, num_warps=1, num_stages=1),
+        triton.Config({"SIZE_N": 1}, num_warps=1, num_stages=1),
+    ],
+    key=["vec_cols", "mat_rows", "mat_cols", "out_cols"],
+)
+@triton.jit
+def vec_mat(
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/implementations/attention_vec_mat.py:23]: done

File: src/kernl/implementations/attention_vec_mat.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        mat_rows, mat_cols = mat_cols, mat_rows
+
+    assert vec.shape[-2] == output.shape[-2] == 1
+    assert mat_cols == out_cols
+    assert vec_cols == mat_rows
+
+    def grid(args) -> Tuple[int, int, int]:
+        return triton.cdiv(mat_cols, args["SIZE_N"]), heads, batch
+
+    vec_cols_pow_2 = triton.next_power_of_2(vec_cols)
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/implementations/attention_vec_mat.py:132]: open for a proposition

File: experimental/misc/dot_perf.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            out,
+            *vec.stride(),
+            *matrix.stride(),
+            *out.stride(),
+            size_m,
+            d_head,
+            use_dot,
+        )
+        end_event[i].record()
+    times_overhead = torch.median(torch.tensor([s.elapsed_time(e) for s, e in zip(start_event, end_event)]))
```
Comment (gaetansnl) [inline experimental/misc/dot_perf.py:121]: don't need sync here ?

**Review by gaetansnl — APPROVED** [review]
just added comment about cuda synchronization, not sure

File: experimental/misc/dot_perf.py (hunk 8 above)
Comment (pommedeterresautee) [reply experimental/misc/dot_perf.py:121]: It appears you that you are right, but again, what is the purpose of this cuda time if it measures with/without sync do not match...
